### PR TITLE
[ADD] mass_mailing_list_dynamic: Autopopulated mailing lists

### PR DESCRIPTION
**Pull request by yajo**
# Dynamic Mass Mailing Lists

Without this addon you have to choose between providing a dynamic domain and
letting your mass mailings reach all partners that match it, or being able to
unsubscribe to certain mailing lists while still being subscribed to others.

This addon allows you to create dynamic mailing lists, so you can now benefit
from both things.

Usage
=====

To create a dynamic mailing list, you need to:

1. Go to *Mass Mailing > Mailings > Mailing Lists > Create*.
1. Choose a *Sync method*.
1. Choose a *Sync domain*.

When you hit the *Sync now* button or send a mass mailing to this list, its
contacts will be automatically updated.

Pay attention to the messages shown to you that tell you about some non-obvious
behaviour you could experience if you edit manually contacts from a dynamic
list.

---

Affected by https://github.com/odoo/odoo/issues/20603
@Tecnativa

### CODE REVIEW

File: mass_mailing_list_dynamic/README.rst [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+This addon allows you to create dynamic mailing lists, so you can now benefit
+from both things.
+
+Usage
+=====
+
+To create a dynamic mailing list, you need to:
+
+#. Go to *Mass Mailing > Mailings > Mailing Lists > Create*.
```
Comment (pedrobaeza) [inline mass_mailing_list_dynamic/README.rst]: Put this on configuration section. The "Sync now" paragraph and following can be in the usage section.

File: mass_mailing_list_dynamic/README.rst [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+#. Choose a *Sync method*.
```
Comment (pedrobaeza) [inline mass_mailing_list_dynamic/README.rst]: Explain what are the possible options.

File: mass_mailing_list_dynamic/README.rst [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+#. Choose a *Sync domain*.
```
Comment (pedrobaeza) [inline mass_mailing_list_dynamic/README.rst]: Explain what is a domain. Domain is not a user word. It shouldn't even go with that word in the interface. Look at mass mailings domain field how is called.

File: mass_mailing_list_dynamic/README.rst [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+Images
+------
+
+* Odoo Community Association: `Icon <https://github.com/OCA/maintainer-tools/blob/master/template/module/static/description/icon.svg>`_.
+
+Contributors
+------------
+
+* Jairo Llopis <jairo.llopis@tecnativa.com> (`Tecnativa <https://www.tecnativa.com>`_)
```
Comment (pedrobaeza) [inline mass_mailing_list_dynamic/README.rst]: Better to put:
```
* Tecnativa (https://www.tecnativa.com):
  * Jairo Llopis <jairo.llopis@tecnativa.com>
```
Propose this to the template if you think it's needed.

File: mass_mailing_list_dynamic/__manifest__.py [hunk 5]
```diff
@@ -0,0 +1,21 @@
+# -*- coding: utf-8 -*-
+# Copyright 2017 Jairo Llopis <jairo.llopis@tecnativa.com>
```
Comment (pedrobaeza) [inline mass_mailing_list_dynamic/__manifest__.py]: Don't add mail to the rest of the files, as it's already in the README:
```python
# Copyright 2017 Tecnativa - Jairo Llopis
```

File: mass_mailing_list_dynamic/__manifest__.py [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+{
+    "name": "Dynamic Mass Mailing Lists",
+    "summary": "Mass mailing lists that get autopopulated",
+    "version": "10.0.1.0.0",
+    "category": "Marketing",
+    "website":
+        "https://github.com/OCA/social/tree/10.0/mass_mailing_list_dynamic",
```
Comment (pedrobaeza) [inline mass_mailing_list_dynamic/__manifest__.py]: This should be main URL: `https://github.com/OCA/social`

File: mass_mailing_list_dynamic/i18n/es.po [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+#. module: mass_mailing_list_dynamic
+#: selection:mail.mass_mailing.list,sync_method:0
+msgid "Only add new records"
+msgstr "Solamente añadir nuevos registros"
+
+#. module: mass_mailing_list_dynamic
+#: model:ir.model.fields,field_description:mass_mailing_list_dynamic.field_mail_mass_mailing_list_sync_domain
+msgid "Sync domain"
+msgstr "Dominio de sincronización"
```
Comment (pedrobaeza) [inline mass_mailing_list_dynamic/i18n/es.po]: We can put `Synchronization criteria` / `Criterios para sincronización`

File: mass_mailing_list_dynamic/i18n/es.po [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+
+#. module: mass_mailing_list_dynamic
+#: model:ir.model.fields,field_description:mass_mailing_list_dynamic.field_mail_mass_mailing_list_sync_method
+msgid "Sync method"
```
Comment (pedrobaeza) [inline mass_mailing_list_dynamic/i18n/es.po]: Don't abbreviate: `Syncronization method`

File: mass_mailing_list_dynamic/models/mail_mass_mailing_list.py [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from odoo import api, fields, models
+from odoo.tools import safe_eval
+
+
+class MassMailingList(models.Model):
+    _inherit = "mail.mass_mailing.list"
+
+    sync_method = fields.Selection(
```
Comment (pedrobaeza) [inline mass_mailing_list_dynamic/models/mail_mass_mailing_list.py]: Please add the flag that I specify you: `Dynamic list`. If not marked, the rest of the fields are invisible in the view.

File: mass_mailing_list_dynamic/models/mail_mass_mailing_list.py [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+            for partner in desired_partners - current_partners:
+                Contact.create({
+                    "list_id": one.id,
+                    "partner_id": partner.id,
+                })
+        # Invalidate cached contact count
+        self.invalidate_cache(["contact_nbr"], dynamic.ids)
+
+    # TODO Delete when fixed https://github.com/OCA/oca-decorators/issues/6
+    def write(self, vals):
```
Comment (pedrobaeza) [inline mass_mailing_list_dynamic/models/mail_mass_mailing_list.py]: I don't think we should sync on create/write. Just when manually done or when sending mailings.

File: mass_mailing_list_dynamic/README.rst [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+.. image:: https://odoo-community.org/website/image/ir.attachment/5784_f2813bd/datas
+   :alt: Try me on Runbot
+   :target: https://runbot.odoo-community.org/runbot/205/10.0
+
+Known issues / Roadmap
+======================
+
+* Tests affected by https://github.com/odoo/odoo/issues/20603. Do not run them
+  in stateful databases.
+* Forbid users about adding or removing contacts to fully synchronized lists.
```
Comment (pedrobaeza) [inline mass_mailing_list_dynamic/README.rst]: Please implement this. You can do it intercepting `create` method of a list contact and checking sync_method == full sync plus the flag that is still missing on the owner list. This way, you don't need to make UI tricks and you cover also the programmatic case. This only requires a context key to be checked for allowing to add contacts when running sync method.

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

File: mass_mailing_list_dynamic/README.rst (hunk 3 above)
Comment (yajo) [reply mass_mailing_list_dynamic/README.rst]: It says "Domain" there too. The field includes a *help* attribute that explains its meaning, but I'll add the doc here too.

File: mass_mailing_list_dynamic/models/mail_mass_mailing_list.py (hunk 9 above)
Comment (yajo) [reply mass_mailing_list_dynamic/models/mail_mass_mailing_list.py]: It's not needed, just leave this field empty

Comment (pedrobaeza) [reply mass_mailing_list_dynamic/models/mail_mass_mailing_list.py]: But that's not what I want. Trust me: this is very user unfriendly. Please add the field.

**Comment by yajo** [comment]
It should be ready now.

**Review by rafaelbn — APPROVED** [review]
Tested functionally in runbot 👍

**Comment by yajo** [comment]
Fixed a bug that was deleting all contacts in database 😱

**Review by pedrobaeza — APPROVED** [review]

**Pull request merged**